runtime/lua/vim/ui/img.lua

```diff
@@ -0,0 +1,486 @@
+---@param opts? {write?:fun(...:string)}
+---@return vim.ui.img.Provider|nil
+function M:__load_provider(opts)
+  return require('vim.ui.img.providers').load(self.__provider, opts)
+end
+
+---Displays the image.
+---```lua
+----- Some predefined image
+---local img = vim.ui.img.new({ ... })
+---
+-----Can be invoked synchronously
+---assert(img:show({ ... }):wait())
+---
+-----Can also be invoked asynchronously
+---img:show({ ... }):on_done(function(err)
+---  -- Do something
+---end)
+---```
+---@param opts? vim.ui.img.Opts
+---@return vim.ui.img._Promise<true>
+function M:show(opts)
+  -- Update the opts to reflect what we should be showing next
+  -- such that future updates can work properly in batch
+  self.__opts = require('vim.ui.img.opts').new(opts)
+
+  return self:__schedule(self.__show, self, self.__opts:into_internal_opts())
+end
+
+---@private
+---@param opts vim.ui.img.InternalOpts
+---@return vim.ui.img._Promise<true>
+function M:__show(opts)
+  local promise = require('vim.ui.img._promise').new({
+    context = 'image.show',
+  })
+
+  local provider = self:__load_provider()
+  if not provider then
+    promise:fail('unable to retrieve provider')
+  else
+    provider
+      .show(self, opts)
+      :on_ok(function(id)
+        self.__id = id
+        promise:ok(true)
+      end)
+      :on_fail(function(show_err)
+        promise:fail(show_err)
+      end)
+  end
+
+  return promise
+end
+
+---Hides the image.
+---```lua
+----- Some predefined image
+---local img = vim.ui.img.new({ ... })
+---
+-----Can be invoked synchronously
+---assert(img:hide():wait())
+---
+-----Can also be invoked asynchronously
+---img:hide():on_done(function(err)
+---  -- Do something
+---end)
+---```
+---@return vim.ui.img._Promise<true>
+function M:hide()
+  -- Update the opts to reflect that we are showing nothing next
+  -- such that future updates can work properly in batch
+  self.__opts = nil
+
+  return self:__schedule(self.__hide, self)
+end
+
+---@private
+---@return vim.ui.img._Promise<true>
+function M:__hide()
+  local promise = require('vim.ui.img._promise').new({
+    context = 'image.hide',
+  })
+
+  local provider = self:__load_provider()
+  if not provider then
+    promise:fail('unable to retrieve provider')
+  else
+    provider
+      .hide(self.__id)
+      :on_ok(function()
+        self.__id = nil
+        promise:ok(true)
+      end)
+      :on_fail(function(hide_err)
+        promise:fail(hide_err)
+      end)
+  end
+
+  return promise
+end
+
+---Updates the image by altering any of the specified `opts`.
+---```lua
+----- Some predefined image
+---local img = vim.ui.img.new({ ... })
+---
+-----Can be invoked synchronously
+---assert(img:update({ ... }):wait())
+---
+-----Can also be invoked asynchronously
+---img:update({ ... }):on_done(function(err)
+---  -- Do something
+---end)
+---```
+---@param opts? vim.ui.img.Opts
+---@return vim.ui.img._Promise<true>
+function M:update(opts)
```

Comment:
Update and show work almost the opposite way of what I would expect.

On one hand, show is used for initially showing an image, but on the other hand it can also be thought of a pair to hide. And in that case, I would like to be able to just re-use the existing options without specifying anything.

I think we could use only one function, but support three modes
1. nil opts - keep the existing configuration
2. partial opts - merge the configuration
3. full opts - replace the configuration

The difference between 2. and 3. could for example be expressed with a special field `update=true` to select mode 3, otherwise mode 2. is used.